src/librbd/ImageWatcher.cc

```diff
@@ -121,8 +131,19 @@ int ImageWatcher::unregister_watch() {
 
   cancel_async_requests(-ESHUTDOWN);
 
-  RWLock::WLocker l(m_watch_lock);
-  return m_image_ctx.md_ctx.unwatch2(m_handle);
+  int r = 0;
+  {
+    RWLock::WLocker l(m_watch_lock);
+    assert(m_watch_state != WATCH_STATE_UNREGISTERED);
+    if (m_watch_state == WATCH_STATE_REGISTERED) {
```

Comment:
```reregister_watch``` should only be hit when the watch state is indicated as an error -- so it's probably better if I am more explicit in that method's check.  I want to avoid unwatching in the case of an error since the handle is already invalid.